src/dev/flang/ir/IR.java

```diff
@@ -337,6 +340,204 @@ else if (e instanceof Universe)
   }
 
 
+  /**
+   * Check if expr might need boxing or tagging and wrap this
+   * into Box()/Tag()/Tag(Box()) if this is the case.
+   *
+   * @param expr the expr to be boxed/tagged
+   *
+   * @param frmlT the formal type this value is assigned to
+   *
+   * @return this or an instance of Box/Tag wrapping this.
+   */
+  protected static Expr boxAndTag(Expr expr, AbstractType frmlT)
+  {
+    if (PRECONDITIONS) require
+      (frmlT != null);
+
+    var result = expr;
+    var t = expr.type();
+
+    if (!t.isVoid() && frmlT.isAssignableFrom(t).yes())
+      {
+        var rt = needsBoxing(expr, frmlT);
+        if (rt != null)
+          {
+            result = new Box(result, rt);
+          }
+        if (frmlT.isChoice() && frmlT.isAssignableFrom(result.type()).yes())
+          {
+            result = tag(result, frmlT);
+            if (CHECKS) check
+              (needsBoxing(result, frmlT) == null);
+          }
+      }
+    // NYI: ugly special case: currently needed for code like
+    // because isAssignableFrom does not return yes without correct Context...
```

Comment:
I guess the `code like` refers to the code underneath this comment. Should be rearranged.